### ISSUE TITLE
Group deploy task picked up wrong/deleted imageID iff hostCount==1 && imageID >= 10

### ISSUE THREAD

**Issue by ipaq3870** [FOGProject/fogproject#393]
As titled. Proposed fix:

https://github.com/FOGProject/fogproject/blob/b619bfe7b4cd8afff80a7c0a3fa56c1f6d3741e3/packages/web/lib/fog/group.class.php#L624

$imageIDs[] = self::getSubObjectIDs(...);

**Comment by Sebastian-Roth** [comment]
@ipaq3870 Thanks for bringing this up. May I raise a few questions? What do you mean by "deleted" in the title? Why would this only happen in case of imageID >= 10??

**Comment by ipaq3870** [comment]
Consider the following simplified code snippet:
```
            elseif ($TaskType->isDeploy()) {
                $imageIDs = self::getSubObjectIDs(
                    'Host',
                    array(
                        'id' => $hostIDs,
                    ),
                    'imageID',
                    false,
                    'AND',
                    'name',
                    false,
                    ''
                );
                for ($i = 0; $i < $hostCount; ++$i) {
                    $batchTask[] = array(
                        $taskName,
                        $username,
                        $hostIDs[$i],
                        0,
                        self::getQueuedState(),
                        $TaskType->get('id'),
                        $StorageNode->get('id'),
                        $wol,
                        $imageIDs[$i],
                        $shutdown,
                        $debug,
                        $passreset,
                    );
                }
```
If hostCount = 1, the self::getSubObjectIDs() call above is supposed to return an array of only 1 member, but when assigning the array to the variable $imageIDs which is not declared an array, it looks like PHP will convert the only 1 array member to a string.

Then with the for loop, the statement $batchTask[] = array(... $hostIDs[$i] ...) will pick up $imageIDs[0], which is the first character of the corresponding imageID converted to string, and there you go.

The case that I stumbled upon:
1. deploy task with a group containing only 1 host (hostCount == 1);
2. the image ID associated with the host is 22 (imageID >=10, i.e. imageID consists of 2 or more digits);
3. if imageID==2 does not exists (deleted), result in osID=0 error in FOS;
4. if imageID==2 actually does exist, it will get deployed wrongly instead of imageID=22.

The fix proposed explicitly defined $imageIDs[] as an array to avoid the problem above mentioned.

P.S. edited subject line to refer to "wrong/deleted imageID" instead of "wrong/deleted image" for clarify.

**Comment by Sebastian-Roth** [comment]
@ipaq3870 Thanks heaps for explaining this case more thoroughly. Now I see what you mean and can confirm this being an issue. I suppose we never ran into this before as deploying to a **group with only one member** is seldomly used. But sure you are right.

**Comment by Sebastian-Roth** [comment]
Done

**Comment by Sebastian-Roth** [comment]
Turns out this is causing some major trouble:
https://forums.fogproject.org/topic/14703/weird-problem-when-deploying-to-groups
https://forums.fogproject.org/topic/14702/invalid-os-id-0-determineos-args-passed-0

Will look into this soon!

**Comment by Sebastian-Roth** [comment]
@ipaq3870 After looking into this closer I found that the conversion from Array to String does not happen "by accident" but is intended: https://github.com/FOGProject/fogproject/blob/master/packages/web/lib/fog/fogmanagercontroller.class.php#L410

I am sure this is for a good reason and I won't touch that part of the code because way to many other parts call this and I am sure it will break things. So I might just add a `is_array` check in the group.class.php code instead of making it an array by default.

**Comment by Sebastian-Roth** [comment]
Fixed in e4f1993
